app/src/main/java/com/keylesspalace/tusky/util/SmartLengthInputFilter.java

```diff
@@ -0,0 +1,145 @@
+ * constraints and adds better visuals such as:
+ * <ul>
+ *     <li>Ellipsis at the end of the constrained text to show continuation.</li>
+ *     <li>Trimming of invisible characters (new lines, spaces, etc.) from the constrained text.</li>
+ *     <li>Constraints end at the end of the last "word", before a whitespace.</li>
+ *     <li>Expansion of the limit by up to 10 characters to facilitate the previous constraint.</li>
+ *     <li>Constraints are not applied if the percentage of hidden content is too small.</li>
+ * </ul>
+ *
+ * Some of these features are configurable through at instancing time.
+ */
+public class SmartLengthInputFilter implements InputFilter {
+
+	/**
+	 * Default for maximum status length on Mastodon and default collapsing
+	 * length on Pleroma.
+	 */
+	public static final int LENGTH_DEFAULT = 500;
+
+	private final int max;
+	private final boolean allowRunway;
+	private final boolean skipIfBadRatio;
+
+	/**
+	 * Creates a new {@link SmartLengthInputFilter} instance with a predefined maximum length and
+	 * all the smart constraint features this class supports.
+	 *
+	 * @param max The maximum length before trimming. May change based on other constraints.
+	 */
+	public SmartLengthInputFilter(int max) {
+		this(max, true, true);
+	}
+
+	/**
+	 * Fully configures a new {@link SmartLengthInputFilter} to fine tune the state of the
+	 * supported smart constraints this class supports.
+	 *
+	 * @param max            The maximum length before trimming.
+	 * @param allowRunway    Whether to extend {@param max} by an extra 10 characters
+	 *                       and trim precisely at the end of the closest word.
+	 * @param skipIfBadRatio Whether to skip trimming entirely if the trimmed content
+	 *                       will be less than 25% of the shown content.
+	 */
+	public SmartLengthInputFilter(int max, boolean allowRunway, boolean skipIfBadRatio) {
+		this.max = max;
+		this.allowRunway = allowRunway;
+		this.skipIfBadRatio = skipIfBadRatio;
+	}
+
+	/**
+	 * Calculates if it's worth trimming the message at a specific limit or if the content
+	 * that will be hidden will not be enough to justify the operation.
+	 *
+	 * @param message The message to trim.
+	 * @param limit   The maximum length after trimming.
+	 * @return        Whether the message should be trimmed or not.
+	 */
+	public static boolean hasBadRatio(Spanned message, int limit) {
+		return (double) limit / message.length() > 0.75;
+	}
+
+	/** {@inheritDoc} */
+	@Override
+	public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
+		// Code originally imported from InputFilter.LengthFilter but heavily customized.
+		// https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/text/InputFilter.java#175
+
+		int sourceLength = source.length();
+		int keep = max - (dest.length() - (dend - dstart));
+		if(keep <= 0) return "";
```

Comment:
Not sure if we need to check that but can be left for keeping it generic I guess
and put spaces after `if`s please

Reply:
That was actually checked in Google's original `LengthFilter` code. I left it in as it catches one more particular case without having to go through the logic below which would become quite expensive in comparison.

I didn't notice you styled if blocks spaced, I may have to go through all of the code I touched and manually put them in if that's the case because I don't have that set in my style settings.